From 1e325362dc0f7febf48218d4b29a92bdc87e4d67 Mon Sep 17 00:00:00 2001
From: Joshua Wilkinson <joshua.wilkinson@morsemicro.com>
Date: Wed, 30 Apr 2025 07:22:31 +0000
Subject: [PATCH 5/7] Merged in SW-14454-add-PKEX-commit-reveal-retries (pull
 request #284)
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

SW-14454 add PKEX Commit Reveal Request Retries

Add the ability for hostapd to retry the PKEX Commit-Reveal Request message up to 5 times after not receiving a PKEX Commit-Reveal Response for 2s.

Note: If the supplicant rx’s an Exchange Request retry after it sent an Exchange Response, it will currently ignore the new request. In this change I only ignore the request if the peer mac address does not match, since it is possible for the Exchange Response to be lost (and this is not retried).

Approved-by: Andrew Pope
Approved-by: Nev Young
Approved-by: Simon Wadsworth
Approved-by: Adam Baumann
Approved-by: Sam Chapman
Approved-by: James Herbert
---
 src/ap/dpp_hostapd.c            | 23 ++++++++-
 src/common/dpp.h                |  2 +
 src/common/dpp_pkex.c           |  1 +
 wpa_supplicant/dpp_supplicant.c | 82 +++++++++++++++++----------------
 4 files changed, 66 insertions(+), 42 deletions(-)

diff --git a/src/ap/dpp_hostapd.c b/src/ap/dpp_hostapd.c
index 1556b1ffd..1ccff4e37 100644
--- a/src/ap/dpp_hostapd.c
+++ b/src/ap/dpp_hostapd.c
@@ -470,7 +470,8 @@ static void hostapd_dpp_pkex_retry_timeout(void *eloop_ctx, void *timeout_ctx)
 
 	if (!pkex || !pkex->exchange_req)
 		return;
-	if (pkex->exch_req_tries >= 5) {
+	if ((!pkex->exchange_done && pkex->exch_req_tries >= 5) ||
+	     (pkex->exchange_done && pkex->commit_reveal_tries >= 5)) {
 		if (hostapd_dpp_pkex_next_channel(hapd, pkex) < 0) {
 #ifdef CONFIG_DPP3
 			if (pkex->v2 && !pkex->forced_ver) {
@@ -488,6 +489,17 @@ static void hostapd_dpp_pkex_retry_timeout(void *eloop_ctx, void *timeout_ctx)
 			return;
 		}
 		pkex->exch_req_tries = 0;
+		pkex->commit_reveal_tries = 0;
+	}
+
+	if (pkex->exchange_done) {
+		pkex->commit_reveal_tries++;
+		wpa_printf(MSG_DEBUG, "DPP: Retransmit PKEX Commit-Reveal Request (try %d)",
+			   pkex->commit_reveal_tries);
+		hostapd_drv_send_action(hapd, pkex->freq, 0, pkex->peer_mac,
+					wpabuf_head(pkex->commit_reveal_req),
+					wpabuf_len(pkex->commit_reveal_req));
+		return;
 	}
 
 	pkex->exch_req_tries++;
@@ -520,6 +532,12 @@ static void hostapd_dpp_pkex_tx_status(struct hostapd_data *hapd, const u8 *dst,
 		return;
 	}
 
+	if (hapd->dpp_pkex->exchange_done) {
+		eloop_cancel_timeout(hostapd_dpp_pkex_retry_timeout, hapd, NULL);
+		eloop_register_timeout(2, 0, hostapd_dpp_pkex_retry_timeout, hapd, NULL);
+		return;
+	}
+
 	if (pkex->exch_req_wait_time && pkex->exchange_req) {
 		/* Wait for PKEX Exchange Response frame and retry request if
 		 * no response is seen. */
@@ -2318,6 +2336,7 @@ hostapd_dpp_rx_pkex_exchange_resp(struct hostapd_data *hapd, const u8 *src,
 	hapd->dpp_pkex->exch_req_wait_time = 0;
 
 	msg = dpp_pkex_rx_exchange_resp(hapd->dpp_pkex, src, buf, len);
+	hapd->dpp_pkex->commit_reveal_req = msg;
 	if (!msg) {
 		wpa_printf(MSG_DEBUG, "DPP: Failed to process the response");
 		return;
@@ -2331,7 +2350,6 @@ hostapd_dpp_rx_pkex_exchange_resp(struct hostapd_data *hapd, const u8 *src,
 		DPP_PA_PKEX_COMMIT_REVEAL_REQ);
 	hostapd_drv_send_action(hapd, freq, 0, src,
 				wpabuf_head(msg), wpabuf_len(msg));
-	wpabuf_free(msg);
 }
 
 
@@ -2395,6 +2413,7 @@ hostapd_dpp_rx_pkex_commit_reveal_resp(struct hostapd_data *hapd, const u8 *src,
 	struct dpp_pkex *pkex = hapd->dpp_pkex;
 	char cmd[500];
 
+	eloop_cancel_timeout(hostapd_dpp_pkex_retry_timeout, hapd, NULL);
 	wpa_printf(MSG_DEBUG, "DPP: PKEX Commit-Reveal Response from " MACSTR,
 		   MAC2STR(src));
 
diff --git a/src/common/dpp.h b/src/common/dpp.h
index 159193abc..e29f20952 100644
--- a/src/common/dpp.h
+++ b/src/common/dpp.h
@@ -224,9 +224,11 @@ struct dpp_pkex {
 	struct crypto_ec_key *peer_bootstrap_key;
 	struct wpabuf *exchange_req;
 	struct wpabuf *exchange_resp;
+	struct wpabuf *commit_reveal_req;
 	unsigned int t; /* number of failures on code use */
 	unsigned int exch_req_wait_time;
 	unsigned int exch_req_tries;
+	unsigned int commit_reveal_tries;
 	unsigned int freq;
 	u8 peer_version;
 	struct wpabuf *enc_key;
diff --git a/src/common/dpp_pkex.c b/src/common/dpp_pkex.c
index dca0d8d39..057f28709 100644
--- a/src/common/dpp_pkex.c
+++ b/src/common/dpp_pkex.c
@@ -1383,6 +1383,7 @@ void dpp_pkex_free(struct dpp_pkex *pkex)
 	crypto_ec_key_deinit(pkex->peer_bootstrap_key);
 	wpabuf_free(pkex->exchange_req);
 	wpabuf_free(pkex->exchange_resp);
+	wpabuf_free(pkex->commit_reveal_req);
 	wpabuf_free(pkex->enc_key);
 	os_free(pkex);
 }
diff --git a/wpa_supplicant/dpp_supplicant.c b/wpa_supplicant/dpp_supplicant.c
index cdcd0f1de..ce4bc1f54 100644
--- a/wpa_supplicant/dpp_supplicant.c
+++ b/wpa_supplicant/dpp_supplicant.c
@@ -3189,6 +3189,44 @@ static void wpas_dpp_pkex_retry_timeout(void *eloop_ctx, void *timeout_ctx)
 			       wpas_dpp_tx_pkex_status, 0);
 }
 
+static struct dpp_bootstrap_info *
+wpas_dpp_pkex_finish(struct wpa_supplicant *wpa_s, const u8 *peer,
+		     unsigned int freq)
+{
+	struct dpp_bootstrap_info *bi;
+
+	wpas_dpp_pkex_clear_code(wpa_s);
+	bi = dpp_pkex_finish(wpa_s->dpp, wpa_s->dpp_pkex, peer, freq);
+	if (!bi)
+		return NULL;
+
+	wpa_s->dpp_pkex = NULL;
+
+#ifdef CONFIG_DPP3
+	if (wpa_s->dpp_pb_bi && !wpa_s->dpp_pb_configurator &&
+	    os_memcmp(bi->pubkey_hash_chirp, wpa_s->dpp_pb_init_hash,
+		      SHA256_MAC_LEN) != 0) {
+		char id[20];
+
+		wpa_printf(MSG_INFO,
+			   "DPP: Peer bootstrap key from PKEX does not match PB announcement response hash");
+		wpa_hexdump(MSG_DEBUG,
+			    "DPP: Peer provided bootstrap key hash(chirp) from PB PKEX",
+			    bi->pubkey_hash_chirp, SHA256_MAC_LEN);
+		wpa_hexdump(MSG_DEBUG,
+			    "DPP: Peer provided bootstrap key hash(chirp) from PB announcement response",
+			    wpa_s->dpp_pb_init_hash, SHA256_MAC_LEN);
+
+		os_snprintf(id, sizeof(id), "%u", bi->id);
+		dpp_bootstrap_remove(wpa_s->dpp, id);
+		wpas_dpp_push_button_stop(wpa_s);
+		return NULL;
+	}
+#endif /* CONFIG_DPP3 */
+
+	return bi;
+}
+
 
 static void
 wpas_dpp_tx_pkex_status(struct wpa_supplicant *wpa_s,
@@ -3234,6 +3272,9 @@ wpas_dpp_tx_pkex_status(struct wpa_supplicant *wpa_s,
 				       wpas_dpp_pkex_retry_timeout, wpa_s,
 				       NULL);
 	}
+
+	if (result == OFFCHANNEL_SEND_ACTION_SUCCESS && wpa_s->dpp_pkex_wait_auth_req)
+		wpas_dpp_pkex_finish(wpa_s, src, freq);
 }
 
 
@@ -3276,7 +3317,7 @@ wpas_dpp_rx_pkex_exchange_req(struct wpa_supplicant *wpa_s, const u8 *src,
 	}
 #endif /* CONFIG_DPP2 */
 
-	if (wpa_s->dpp_pkex) {
+	if (wpa_s->dpp_pkex && memcmp(src, wpa_s->dpp_pkex->peer_mac, ETH_ALEN)) {
 		/* TODO: Support parallel operations */
 		wpa_printf(MSG_DEBUG,
 			   "DPP: Already in PKEX session - ignore new request");
@@ -3361,44 +3402,6 @@ wpas_dpp_rx_pkex_exchange_resp(struct wpa_supplicant *wpa_s, const u8 *src,
 }
 
 
-static struct dpp_bootstrap_info *
-wpas_dpp_pkex_finish(struct wpa_supplicant *wpa_s, const u8 *peer,
-		     unsigned int freq)
-{
-	struct dpp_bootstrap_info *bi;
-
-	wpas_dpp_pkex_clear_code(wpa_s);
-	bi = dpp_pkex_finish(wpa_s->dpp, wpa_s->dpp_pkex, peer, freq);
-	if (!bi)
-		return NULL;
-
-	wpa_s->dpp_pkex = NULL;
-
-#ifdef CONFIG_DPP3
-	if (wpa_s->dpp_pb_bi && !wpa_s->dpp_pb_configurator &&
-	    os_memcmp(bi->pubkey_hash_chirp, wpa_s->dpp_pb_init_hash,
-		      SHA256_MAC_LEN) != 0) {
-		char id[20];
-
-		wpa_printf(MSG_INFO,
-			   "DPP: Peer bootstrap key from PKEX does not match PB announcement response hash");
-		wpa_hexdump(MSG_DEBUG,
-			    "DPP: Peer provided bootstrap key hash(chirp) from PB PKEX",
-			    bi->pubkey_hash_chirp, SHA256_MAC_LEN);
-		wpa_hexdump(MSG_DEBUG,
-			    "DPP: Peer provided bootstrap key hash(chirp) from PB announcement response",
-			    wpa_s->dpp_pb_init_hash, SHA256_MAC_LEN);
-
-		os_snprintf(id, sizeof(id), "%u", bi->id);
-		dpp_bootstrap_remove(wpa_s->dpp, id);
-		wpas_dpp_push_button_stop(wpa_s);
-		return NULL;
-	}
-#endif /* CONFIG_DPP3 */
-
-	return bi;
-}
-
 
 static void
 wpas_dpp_rx_pkex_commit_reveal_req(struct wpa_supplicant *wpa_s, const u8 *src,
@@ -3444,7 +3447,6 @@ wpas_dpp_rx_pkex_commit_reveal_req(struct wpa_supplicant *wpa_s, const u8 *src,
 			       wait_time, wpas_dpp_tx_pkex_status, 0);
 	wpabuf_free(msg);
 
-	wpas_dpp_pkex_finish(wpa_s, src, freq);
 	wpa_s->dpp_pkex_wait_auth_req = true;
 }
 
-- 
2.34.1

